ekm/remote_key_manager.go

```diff
@@ -0,0 +1,455 @@
+		return km.client.AddValidators(context.Background(), arg.(ssvsignerclient.ShareKeys)) // TODO: use context
+	}
+
+	res, err := km.retryFunc(f, shareKeys, "AddValidators")
+	if err != nil {
+		return fmt.Errorf("add validator: %w", err)
+	}
+
+	statuses, ok := res.([]ssvsignerclient.Status)
+	if !ok {
+		return fmt.Errorf("bug: expected []Status, got %T", res)
+	}
+
+	if len(statuses) != 1 {
+		return fmt.Errorf("bug: expected 1 status, got %d", len(statuses))
+	}
+
+	if statuses[0] != ssvsignerclient.StatusImported {
+		return fmt.Errorf("unexpected status %s", statuses[0])
+	}
+
+	if err := km.BumpSlashingProtection(sharePubKey); err != nil {
+		return fmt.Errorf("could not bump slashing protection: %w", err)
+	}
+
+	return nil
+}
+
+func (km *RemoteKeyManager) RemoveShare(pubKey []byte) error {
+	statuses, err := km.client.RemoveValidators(context.Background(), pubKey) // TODO: use context
+	if err != nil {
+		return fmt.Errorf("remove validator: %w", err)
+	}
+
+	if statuses[0] != ssvsignerclient.StatusDeleted {
+		return fmt.Errorf("received status %s", statuses[0])
+	}
+
+	if err := km.RemoveHighestAttestation(pubKey); err != nil {
+		return fmt.Errorf("could not remove highest attestation: %w", err)
+	}
+	if err := km.RemoveHighestProposal(pubKey); err != nil {
+		return fmt.Errorf("could not remove highest proposal: %w", err)
+	}
+
+	return nil
+}
+
+func (km *RemoteKeyManager) retryFunc(f func(arg any) (any, error), arg any, funcName string) (any, error) {
+	if km.retryCount < 2 {
+		return f(arg)
+	}
+
+	var multiErr error
+	for i := 1; i <= km.retryCount; i++ {
+		v, err := f(arg)
+		if err == nil {
+			return v, nil
+		}
+		var shareDecryptionError ssvsignerclient.ShareDecryptionError
+		if errors.As(err, &shareDecryptionError) {
+			return nil, ShareDecryptionError(err)
+		}
+		multiErr = errors.Join(multiErr, err)
+		km.logger.Warn("call failed", zap.Error(err), zap.Int("attempt", i), zap.String("func", funcName))
+	}
+
+	return nil, fmt.Errorf("no successful result after %d attempts: %w", km.retryCount, multiErr)
+}
+
+func (km *RemoteKeyManager) SignBeaconObject(
+	obj ssz.HashRoot,
+	domain phase0.Domain,
+	sharePubkey []byte,
+	signatureDomain phase0.DomainType,
+) (spectypes.Signature, [32]byte, error) {
+	forkInfo, err := km.getForkInfo(context.Background()) // TODO: consider passing context to SignBeaconObject
+	if err != nil {
+		return spectypes.Signature{}, [32]byte{}, fmt.Errorf("get fork info: %w", err)
+	}
+
+	req := web3signer.SignRequest{
+		ForkInfo: forkInfo,
+	}
+
+	switch signatureDomain {
+	case spectypes.DomainAttester:
+		data, ok := obj.(*phase0.AttestationData)
+		if !ok {
+			return nil, [32]byte{}, errors.New("could not cast obj to AttestationData")
+		}
+
+		if err := km.IsAttestationSlashable(sharePubkey, data); err != nil {
+			return nil, [32]byte{}, err
+		}
+
+		if err := km.UpdateHighestAttestation(sharePubkey, data); err != nil {
+			return nil, [32]byte{}, err
+		}
```

Comment:
shouldn't we do this only after successful signing?

Reply:
in ekm we don't https://github.com/ssvlabs/eth2-key-manager/blob/v1.5.1/signer/sign_attestation.go#L43

Reply:
I guess this is how we do with local signing as well so this is ok.